configurationapi/oauth2.go

```diff
@@ -0,0 +1,110 @@
+package configurationapi
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+var (
+	_ oauth2.TokenSource = OAuthValues{}
+)
+
+type OAuthValues struct {
+	Transport    *http.Transport
+	TokenUrl     string
+	ClientId     string
+	ClientSecret string
+	Scopes       []string
+}
+
+func (oav OAuthValues) Token() (*oauth2.Token, error) {
+	tokenVal, tokenErr := getToken(oav, context.Background())
+	return tokenVal, tokenErr
+}
+
+func getToken(oav OAuthValues, ctx context.Context) (*oauth2.Token, error) {
+	var token *oauth2.Token
+	var err error
+
+	oauthConfig := &clientcredentials.Config{
+		TokenURL:     oav.TokenUrl,
+		ClientID:     oav.ClientId,
+		ClientSecret: oav.ClientSecret,
+		Scopes:       oav.Scopes,
+	}
+
+	httpClient := &http.Client{Transport: oav.Transport}
+	ctx = context.WithValue(ctx, oauth2.HTTPClient, httpClient)
+
+	token, err = oAuthExponentialBackOffRetry(func() (*oauth2.Token, error) {
+		return oAuthProcessResponse(func() (*oauth2.Token, error) {
+			return oauthConfig.Token(ctx)
+		})
+	})
+
+	if err != nil {
```

Comment:
You should be able to just return directly from line 46, these checks won't change what gets returned